### ISSUE TITLE
why 'end_points['fp2_inds'] = end_points['sa1_inds'][:, 0:num_seed]' ?

### ISSUE THREAD

**Issue by linhaojia13** [zeliu98/Group-Free-3D#31]
In the models/backbone_module.py, you select the first 1024 out of 2048 `sa1_inds` as `fp2_inds`. I can understand that the intention behind this is to obtain the indices of these 1024 seed points in the entire point cloud, in order to participate in the loss calculation in the function `compute_points_obj_cls_loss_hard_topk`. 

However, directly selecting the first 1024 out of 2048 `sa1_inds` does not correspond one-to-one with `fp2_xyz`. This mismatch would cause `euclidean_dist1` and `object_assignment_one_hot variables` in the function `compute_points_obj_cls_loss_hard_topk` to not be aligned one-to-one. Doesn't this introduce an error in the supervision signal for KPS?

models/backbone_module.py:
```
        # --------- 2 FEATURE UPSAMPLING LAYERS --------
        features = self.fp1(end_points['sa3_xyz'], end_points['sa4_xyz'], end_points['sa3_features'],
                            end_points['sa4_features'])
        features = self.fp2(end_points['sa2_xyz'], end_points['sa3_xyz'], end_points['sa2_features'], features)
        end_points['fp2_features'] = features
        end_points['fp2_xyz'] = end_points['sa2_xyz']
        num_seed = end_points['fp2_xyz'].shape[1]
        end_points['fp2_inds'] = end_points['sa1_inds'][:, 0:num_seed]  # indices among the entire input point clouds

        return end_points
```

models/loss_helper.py:
```
def compute_points_obj_cls_loss_hard_topk(end_points, topk):
    box_label_mask = end_points['box_label_mask']
    seed_inds = end_points['seed_inds'].long()  # B, K
    seed_xyz = end_points['seed_xyz']  # B, K, 3
    seeds_obj_cls_logits = end_points['seeds_obj_cls_logits']  # B, 1, K
    gt_center = end_points['center_label'][:, :, 0:3]  # B, K2, 3
    gt_size = end_points['size_gts'][:, :, 0:3]  # B, K2, 3
    B = gt_center.shape[0]
    K = seed_xyz.shape[1]
    K2 = gt_center.shape[1]

    point_instance_label = end_points['point_instance_label']  # B, num_points
    object_assignment = torch.gather(point_instance_label, 1, seed_inds)  # B, num_seed
    object_assignment[object_assignment < 0] = K2 - 1  # set background points to the last gt bbox
    object_assignment_one_hot = torch.zeros((B, K, K2)).to(seed_xyz.device)
    object_assignment_one_hot.scatter_(2, object_assignment.unsqueeze(-1), 1)  # (B, K, K2)
    delta_xyz = seed_xyz.unsqueeze(2) - gt_center.unsqueeze(1)  # (B, K, K2, 3)
    delta_xyz = delta_xyz / (gt_size.unsqueeze(1) + 1e-6)  # (B, K, K2, 3)
    new_dist = torch.sum(delta_xyz ** 2, dim=-1)
    euclidean_dist1 = torch.sqrt(new_dist + 1e-6)  # BxKxK2
    euclidean_dist1 = euclidean_dist1 * object_assignment_one_hot + 100 * (1 - object_assignment_one_hot)  # BxKxK2
```

**Comment by L-Reichardt** [comment]
Answering late because I've asked myself the same question.

This behavior stems from the implementation details of the CUDA kernel used for Farthest Point Sampling (FPS). The kernel always selects the first point in the input array as the starting point. For simplicity, consider the point cloud as having shape (N, 3) representing coordinates \[x, y, z].

The input point cloud is initially unstructured, meaning the points are in an arbitrary order. During the first Set Abstraction layer (SA1), the FPS starts at Input\[0, :]. Although the output indices from FPS are not sorted, when these indices are used to gather points, the gathered points become ordered by their relative distances. This means the point at position Input\[q+1, :] is the farthest point relative to Input\[q, :], effectively creating a sorted sequence based on distance.

In subsequent Set Abstraction layers (SA2 and beyond), FPS again starts from the first point of the already ordered set. Since the input to these layers is already ordered by relative distance, the FPS output indices become a simple sequence 0, 1, 2, ..., n-1 corresponding to the number of samples. Therefore, instead of performing FPS repeatedly on an already ordered set, slicing the first n points achieves the same effect.

This explains why slicing the indices obtained from **SA1 works as a proxy sorting for FPS in later layers**. However, slicing the original input point cloud directly would not produce the same result because the initial input is unordered.

The equivalence can be verified with a check such as torch.allclose(sa1\[:npoints\_sa2], sa2), which returns True.

**Other FPS implementations might not work like this**
